File: app/Quotes/bs/src/pages/servers.ml
(*  Copyright 30-Jul-2018 ºDeme
    GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

open Ui
(*open I18n*)

let run = Main.run

let send rq f = Client.send (Main.client ()) rq f

let wrq = Client.wrq

let rrp = Client.rrp

let go name_server = Json.(
  let rq = wrq [
      "page", wstring "servers";
      "rq", wstring "setSelServer";
      "name", wstring name_server
    ]
  in
  send rq (fun _ -> run ())
)

let nick_list server nicks =
  let mk_td nick =
    let field = Wserver_id.mk Nick.(nick.id) server in
    q "td" [Klass "frame"; Style "text-align:center;width:25%;"][
      Wrule.mk_small Nick.(nick.name);
      Wserver_id.widget field]
  in
  let cols = 4 in
  let len = Array.length nicks in
  let rows = 1 + (len - 1) / cols in
  q "table" [Att ("width", "100%"); Klass ("frame")]
    (It.to_list (It.map (fun r ->
    q "tr" []
      (It.to_list (It.map (fun c ->
      let i = c * rows + r in
      if i >= len then
        q "td" [Att ("width", "25%")][]
      else
        mk_td nicks.(i)
      ) (It.range 0 cols)))
    ) (It.range 0 rows)))

let show' menu servers sel nicks =
  let entry name = link (fun _ -> go name) |> Domo.set [
    Klass (if name = sel then "frame" else "link"); Html name]
  and separator () = q "span" [Html " . "][] in
  let w =
    q "div" [][
      q "table" [Att ("align", "center")][
        q "tr" [][
          q "td" [][
            entry servers.(0);
            separator ();
            entry servers.(1);
            separator ();
            entry servers.(2)]]];
      Wrule.mk_big sel;
      nick_list sel nicks;
    ]
  in (
    Main.show menu w
  )

let show menu = Json.(
  let rq = wrq [
      "page", wstring "servers";
      "rq", wstring "idata"
    ]
  in
  send rq (fun rp ->
    let servers = rrp rp "servers" (rit rstring) |> It.to_array
    and sel_server = rrp rp "selServer" rstring
    and nicks = rrp rp "nicks" (rit Nick.of_json) |> It.to_array
    in
    show' menu servers sel_server nicks
  )
)

File: app/Quotes/js/src/Servers.js
// Copyright 05-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import Main from "./Main.js";
import {_} from "./I18n.js";
import Ui from "./dmjs/Ui.js";
import It from "./dmjs/It.js";
import Nick from "./data/Nick.js";
import Wrule from "./wgs/Wrule.js";
import WserverId from "./wgs/WserverId.js";

const $ = Ui.$;

/** Servers page. */
export default class Servers {
  /**
   * @param {!Main} main Main
   */
  constructor (main) {
    /**
     * @private
     * @type {!Main}
     */
    this._main = main;
  }

  async go (server) {
    const data = {
      "source": "servers",
      "rq": "setSelServer",
      "name": server
    };
    await this._main.client.send(data);
    this._main.run();
  }

  // ____
  // View ------------------------------------------------------------
  // TTTT

  nick_list (sel, nicks, codes) {
    const len = nicks.length;
    if (len === 0) {
      return $("table").att("align", "center")
        .add($("tr")
          .add($("td").klass("frame4").html(_("Without nicks"))));
    }
    const cols = 4;
    const rows = Math.floor((len - 1) / cols) + 1;
    return $("table").style("width:100%")
      .adds([...It.range(rows).map(row =>
        $("tr").adds([...It.range(cols).map(col => {
          const ix = row * cols + col;
          if (ix < len) {
            return $("td").klass("frame").style("text-align:center;width:25%;")
              .add(Wrule.mkSmall(nicks[ix].nick))
              .add(new WserverId(
                this._main, sel, nicks[ix].id, codes[ix]
              ).wg());
          }
          return $("td");
        })])
      )]);
  }

  /** @private */
  show2 (servers, sel, nicks, codes) {
    servers.sort();
    const self = this;
    const entry = (server) => Ui.link(() => {
      self.go(server);
    }).klass(server === sel ? "frame" : "link").html(server);
    const separator = () => $("span").html(" · ");
    const w = $("div")
      .add($("table").att("align", "center")
        .add($("tr")
          .add($("td")
            .add(entry(servers[0]))
            .add(separator())
            .add(entry(servers[1]))
            .add(separator())
            .add(entry(servers[2])))))
      .add(Wrule.mkBig(sel))
      .add(self.nick_list(sel, nicks, codes))

    ;
    this._main.dom.show(Main.serversPageId, w);
  }




  /**
   * @return {Promise}
   */
  async show () {
    const self = this;
    const data = {
      "source": "servers",
      "rq": "idata"
    };
    const rp = await self._main.client.send(data);
    const servers = rp["servers"];
    const selServer = rp["selServer"];
    const nicks = rp["nicks"].map(n => Nick.fromJson(n));
    nicks.sort((n1, n2) => n1.nick.localeCompare(n2.nick));
    const codes = [];

    const total = nicks.length;
    let ix = 0;
    const bar = $("div").att("id", "progressFore");
    const wg = $("div").style("width:250px")
      .add($("div").att("id", "progressBack").add(bar));
    const box = self._main.dom.addWait(wg);
    box.show(true);

    const sh = async it => {
      if (it.has) {
        const data = {
          "source": "servers",
          "rq": "nickCode",
          "serverId": selServer,
          "nickId": it.value.id
        };
        const rp = await self._main.client.send(data);
        codes.push(rp["code"]);
        ++ix;
        bar.setStyle("width", `${Math.floor(ix * 100 / total)}%`);
        sh(it.next);
      } else {
        box.show(false);
        self.show2(servers, selServer, nicks, codes);
      }
    };
    sh(It.from(nicks));
  }
}

